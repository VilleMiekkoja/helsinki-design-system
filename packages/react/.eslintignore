**/*.d.ts
webpack.config.js
.templates
lib/
