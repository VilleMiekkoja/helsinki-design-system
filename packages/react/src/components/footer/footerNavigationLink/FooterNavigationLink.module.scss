@value large-down from "../../../styles/breakpoints.css";
@import "../common";

.item {
  @extend %footerItem;
  font-weight: normal;
  margin-top: var(--spacing-3-xs);
  line-height: var(--lineheight-l);

  // spacing between icon and label
  & > *:nth-child(2) {
    margin-left: var(--spacing-3-xs);
  }
}

.base {
  align-self: unset;
  font-size: var(--fontsize-body-s);
  text-decoration: underline;
  font-weight: 500;
  margin: 0;
}

.icon {
  composes: icon from 'hds-core/lib/components/link/link.css';
}

.verticalAlignSmallOrMediumIcon {
  composes: vertical-align-small-or-medium-icon from 'hds-core/lib/components/link/link.css';
}

.navigation {
  align-items: center;
  font-size: var(--fontsize-body-l);
}

.utility {
  align-items: center;
  font-size: var(--fontsize-body-l);
  width: 248px;
  margin-right: var(--spacing-m);
}

.subItem.subItem {
  font-size: var(--fontsize-body-s);
  line-height: var(--lineheight-l)
}

.subItemIcon {
  align-self: start;
  height: var(--spacing-xs);
  min-width: var(--spacing-xs);
  transform: translateY(4px);
  width: var(--spacing-xs);
}

@media medium-up {
  .utility {
    width: 214px
  }
}

@media large-down {
  .item:not(.base, .navigation, .utility) {
    font-size: var(--fontsize-body-m);
  }
}

@media large-up {
  .utility {
    width: 204px
  }
}

@media x-large-up {
  .utility {
    width: 268px
  }
}