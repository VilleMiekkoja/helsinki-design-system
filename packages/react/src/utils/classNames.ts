import classnames from 'classnames/dedupe';

export default classnames;
