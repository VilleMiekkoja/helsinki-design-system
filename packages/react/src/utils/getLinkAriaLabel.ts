import React from 'react';

export default (
  ariaLabel: string,
  label: React.ReactNode,
  openInNewTabAriaLabel?: string,
  openInExternalDomainAriaLabel?: string,
  external?: boolean,
  openInNewTab?: boolean,
) => {
  let childrenText = ariaLabel || label;
  const newTabText = openInNewTab ? openInNewTabAriaLabel || 'Avautuu uudessa välilehdessä.' : '';
  const externalText = external ? openInExternalDomainAriaLabel || 'Siirtyy toiseen sivustoon.' : '';

  if (childrenText && childrenText.slice(-1) !== '.') {
    childrenText = `${childrenText}.`;
  }

  return [childrenText, newTabText, externalText].filter((text) => text).join(' ');
};
