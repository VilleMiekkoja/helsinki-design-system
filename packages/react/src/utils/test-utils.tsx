import React, { PropsWithChildren } from 'react';

import { Navigation } from '../components/navigation';
import { Footer } from '../components/footer';
import { Header } from '../components/header';

export const NavigationWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Navigation menuToggleAriaLabel="menu" skipTo="#content" skipToContentLabel="Skip to content" title="Foo">
    {children}
  </Navigation>
);

export const FooterWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">{children}</Footer>
);

export const FooterNavigationWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Navigation>{children}</Footer.Navigation>
  </Footer>
);

export const FooterNavigationGroupsWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Navigation>
      <Footer.NavigationGroup>{children}</Footer.NavigationGroup>
    </Footer.Navigation>
  </Footer>
);

export const FooterUtilitiesWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Footer title="Bar">
    <Footer.Utilities>{children}</Footer.Utilities>
  </Footer>
);

export const HeaderNavigationMenuWrapper = ({ children }: PropsWithChildren<Record<string, unknown>>) => (
  <Header>
    <Header.NavigationMenu>{children}</Header.NavigationMenu>
  </Header>
);
