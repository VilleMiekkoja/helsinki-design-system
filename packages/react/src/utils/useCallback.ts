import { EventHandler, useCallback } from 'react';

export const useCallbackIfDefined = (callback) => {
  const handler = (event) => {
    if (callback) callback(event);
  };

  return useCallback(handler, [callback]);
};

export const useEnterOrSpacePressCallback = (callback) => {
  const handler = (event) => {
    if (event.key === 'Enter' || event.key === ' ') callback(event);
  };

  return useCallbackIfDefined(handler);
};

export const withDefaultPrevented = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.preventDefault();
    return callback.call(callback, event);
  };

export const withPropagationStopping = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.stopPropagation();
    return callback.call(callback, event);
  };
