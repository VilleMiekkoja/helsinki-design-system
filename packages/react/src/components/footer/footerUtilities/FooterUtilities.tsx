import React, { cloneElement, isValidElement } from 'react';

// import core base styles
import 'hds-core';
import styles from './FooterUtilities.module.scss';
import classNames from '../../../utils/classNames';

export type FooterUtilitiesProps = {
  /**
   * Description of the navigation links for screen readers.
   */
  ariaLabel?: string;
  /**
   * Children elements to render.
   */
  children: React.ReactNode;
  /**
   * List of Footer.NavigationLink components to display in the social media section.
   */
  soMeLinks?: React.ReactNode[];
  /**
   * Props that will be passed to the native `<section>` element.
   * Can be used to pass aria attributes that describes the SoMe section to screen reader users.
   */
  soMeSectionProps?: React.ComponentPropsWithoutRef<'section'>;
};

export const FooterUtilities = ({ ariaLabel, children, soMeLinks, soMeSectionProps }: FooterUtilitiesProps) => {
  return (
    <div className={styles.utilities} aria-label={ariaLabel}>
      <hr className={styles.divider} aria-hidden />
      <div className={classNames(styles.links, !soMeLinks && styles.widerLinks)}>{children}</div>
      {soMeLinks && (
        <section className={styles.soMe} {...soMeSectionProps}>
          {soMeLinks.map((link) => {
            if (isValidElement(link)) {
              /* Set variant to null just in case user set it. It should be null for SoMelinks so it doesn't mess with the styles. */
              return cloneElement(link, { variant: null });
            }
            return null;
          })}
        </section>
      )}
    </div>
  );
};
