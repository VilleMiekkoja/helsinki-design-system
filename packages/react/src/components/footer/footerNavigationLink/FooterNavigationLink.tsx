import React from 'react';

// import core base styles
import 'hds-core';
import styles from './FooterNavigationLink.module.scss';
import { MergeElementProps } from '../../../common/types';
import classNames from '../../../utils/classNames';
import getLinkAriaLabel from '../../../utils/getLinkAriaLabel';
import { IconAngleRight, IconLinkExternal } from '../../../icons';
import { FooterVariant } from '../Footer.interface';

type ItemProps<Element> = React.PropsWithChildren<{
  /**
   * aria-label for providing detailed information for screen readers about a link.
   */
  ariaLabel?: string;
  /**
   * Element type
   */
  as?: Element;
  /**
   * Boolean indicating whether the link will lead user to external domain.
   */
  external?: boolean;
  /**
   * Icon placed on the left side of the item label
   */
  icon?: React.ReactNode;
  /**
   * Boolean indicating whether the link will open in new tab or not.
   */
  openInNewTab?: boolean;
  /**
   * The aria-label for opening link in a new tab
   */
  openInNewTabAriaLabel?: string;
  /**
   * The aria-label for opening link in an external domain
   */
  openInExternalDomainAriaLabel?: string;
  /**
   * The label for the item.
   */
  label?: string;
  /**
   * Whether the item should be displayed as an sub item in a sitemap item group
   * @internal
   */
  subItem?: boolean;
  /**
   * Internal variant to change styles based on context.
   * @internal
   */
  variant?: FooterVariant.Navigation | FooterVariant.Utility;
}>;

export type FooterNavigationLinkProps<Element extends React.ElementType = 'a'> = MergeElementProps<
  Element,
  ItemProps<Element>
>;

export const FooterNavigationLink = <T extends React.ElementType = 'a'>({
  ariaLabel,
  as,
  className,
  icon,
  external = false,
  openInNewTab = false,
  openInExternalDomainAriaLabel,
  openInNewTabAriaLabel,
  label,
  subItem = false,
  variant,
  ...rest
}: FooterNavigationLinkProps<T>) => {
  const Item: React.ElementType = as;
  const composedAriaLabel = getLinkAriaLabel(
    ariaLabel,
    label,
    openInNewTabAriaLabel,
    openInExternalDomainAriaLabel,
    external,
    openInNewTab,
  );

  return (
    // @ts-ignore
    <Item
      aria-label={ariaLabel}
      className={classNames(styles.item, subItem && styles.subItem, variant && styles[variant], className)}
      external={external}
      openInNewTab={openInNewTab}
      openInExternalDomainAriaLabel={openInExternalDomainAriaLabel}
      openInNewTabAriaLabel={openInNewTabAriaLabel}
      {...(openInNewTab && { target: '_blank', rel: 'noopener' })}
      {...((openInNewTab || external || ariaLabel) && { 'aria-label': composedAriaLabel })}
      {...rest}
    >
      {icon}
      {subItem && <IconAngleRight className={styles.subItemIcon} />}
      {label && <span>{label}</span>}
      {external && (
        <IconLinkExternal
          size="s"
          className={classNames(styles.icon, styles.verticalAlignSmallOrMediumIcon)}
          aria-hidden
        />
      )}
    </Item>
  );
};

FooterNavigationLink.defaultProps = {
  as: 'a',
};
