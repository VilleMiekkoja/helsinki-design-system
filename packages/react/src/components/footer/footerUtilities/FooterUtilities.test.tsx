import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { FooterUtilities } from './FooterUtilities';
import { FooterNavigationLink } from '../footerNavigationLink/FooterNavigationLink';
import { FooterWrapper } from '../../../utils/test-utils';

describe('<Footer.Utilities /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(
      <FooterUtilities>
        <FooterNavigationLink
          href="https://hel.fi/helsinki/fi/kaupunki-ja-hallinto/osallistu-ja-vaikuta/ota-yhteytta/ota-yhteytta"
          label="Yhteystiedot"
        />
        <FooterNavigationLink
          href="https://hel.fi/helsinki/fi/kaupunki-ja-hallinto/osallistu-ja-vaikuta/palaute/anna-palautetta"
          label="Anna ja lue palautetta"
        />
        <FooterNavigationLink href="https://hel.fi/kanslia/neuvonta-fi" label="Chat-neuvonta" />
      </FooterUtilities>,
      { wrapper: FooterWrapper },
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = render(
      <FooterUtilities>
        <FooterNavigationLink
          href="https://hel.fi/helsinki/fi/kaupunki-ja-hallinto/osallistu-ja-vaikuta/ota-yhteytta/ota-yhteytta"
          label="Yhteystiedot"
        />
        <FooterNavigationLink
          href="https://hel.fi/helsinki/fi/kaupunki-ja-hallinto/osallistu-ja-vaikuta/palaute/anna-palautetta"
          label="Anna ja lue palautetta"
        />
        <FooterNavigationLink href="https://hel.fi/kanslia/neuvonta-fi" label="Chat-neuvonta" />
      </FooterUtilities>,
      { wrapper: FooterWrapper },
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
