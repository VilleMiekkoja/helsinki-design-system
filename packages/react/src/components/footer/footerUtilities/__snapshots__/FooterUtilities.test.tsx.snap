// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer.Utilities /> spec renders the component 1`] = `
<DocumentFragment>
  <footer
    class="footer koros-basic theme-light"
  >
    <div
      class="koros basic koros basic"
    >
      <svg
        aria-hidden="true"
        height="85"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <pattern
            height="85"
            id="koros_basic-1"
            patternUnits="userSpaceOnUse"
            width="106"
            x="0"
            y="0"
          >
            <path
              d="M0,800h20V0c-4.9,0-5,2.6-9.9,2.6S5,0,0,0V800z"
              transform="scale(5.3)"
            />
          </pattern>
        </defs>
        <rect
          fill="url(#koros_basic-1)"
          height="85"
          width="100%"
        />
      </svg>
    </div>
    <div
      class="footerContent"
    >
      <section
        class="navigationContainer centeredNavigation"
      >
        <div
          class="titleWrapper"
        >
          <h2
            class="title"
          >
            Bar
          </h2>
        </div>
      </section>
      <div
        class="utilities"
      >
        <hr
          aria-hidden="true"
          class="divider"
        />
        <div
          class="links widerLinks"
        >
          <a
            class="item"
            href="https://hel.fi/helsinki/fi/kaupunki-ja-hallinto/osallistu-ja-vaikuta/ota-yhteytta/ota-yhteytta"
          >
            <span>
              Yhteystiedot
            </span>
          </a>
          <a
            class="item"
            href="https://hel.fi/helsinki/fi/kaupunki-ja-hallinto/osallistu-ja-vaikuta/palaute/anna-palautetta"
          >
            <span>
              Anna ja lue palautetta
            </span>
          </a>
          <a
            class="item"
            href="https://hel.fi/kanslia/neuvonta-fi"
          >
            <span>
              Chat-neuvonta
            </span>
          </a>
        </div>
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
